Geocodificado recursivo

carlosvergara opened the pull request:
Hola de nuevo!

He estado revisando las funciones y parece que los resultados son prácticamente idénticos a la versión previa (PR #3). 

No obstante, me he fijado en que las funciones eran algo engorrosas a la hora de introducirlas dentro de un bucle, de forma que he implementado esta característica como parte de las funciones `cartociudad_geocode()` y `cartociudad_reverse_geocode()`. Concretamente, los cambios son:

1. Sustitución de error en consulta por un aviso + devolución de un `data.frame` con la consulta y valores perdidos para el resto de elementos.
2.  Posibilidad de introducir un vector de direcciones o vectores con latitudes y longitudes, en ambos casos con longitud >= 1.

Un saludo! :)

Commit:
rbind lists to data.frame

Commit:
Multiple queries in the same call

Commit:
Add package purrr to Imports

- Drop rbind_lists_df function.
- Add purrr:map_df() to rbind lists.
- Minor fixes.

Commit:
Better error handling in cartociudad_geocode()

Commit:
Add an initial comprobation

koldLight commented on R/cartociudad_reverse_geocode.R:
```diff
@@ -16,50 +16,49 @@
-    municipio  = info$muni,
-    provincia  = info$province,
-    cod.postal = info$postalCode
-  )
```
Sobre la eliminación de esto, creo que conviene mantener en el paquete el mapeo de nombres, para:

* Mantener coherencia: son nombres con mismo idioma, mismo formato, no como los que devuelve directamente la API (hay inglés, español, underscored, camelCased, ...)
* Mantener retro-compatibilidad con los usuarios actuales
* Asegurar futura compatibilidad: si se lanza una nueva versión de la API y los campos tienen otros nombres, cambiamos el mapeo en el paquete, y la actualización será transparente para los usuarios

Dime cómo lo ves. Si estás de acuerdo, sería bueno mantener el trozo de documentación eliminado sobre los nombres

carlosvergara replied:
Me parece perfecto, aunque creo que sería mejor devolver un data.frame en lugar de una lista. ¿Conservamos esa parte?

koldLight replied:
OK, totalmente de acuerdo, tiene más sentido ahora que está vectorizado

koldLight commented on R/cartociudad_reverse_geocode.R:
```diff
@@ -16,50 +16,49 @@
 #'
+#' # Query multiple points
+#' cartociudad_reverse_geocode(c(40.473219, 39.46979), c(-3.7227241, -0.376963))
+#'
```
Como recomendación, estaría bien añadir un test en `tests/testthat.R` con el caso de llamar a `cartociudad_reverse_geocode` para múltiples localizaciones

carlosvergara replied:
Yo me ocupo de las pruebas.
He pensado en comprobar el funcionamiento ante casos típicos que estoy viendo, como una dirección sin número (devuelve la geometría de toda la vía y la latitud y longitud del portal 0), portales y pk's inexistentes (devolución del punto más próximo), y viales inexistentes.

koldLight replied:
Perfecto!

koldLight commented on R/cartociudad_geocode.R:
```diff
@@ -27,23 +27,54 @@
+#'
+#' # Query multiple addresses
+#' address <- c(address, "plaza del ayunamiento 1, valencia")
+#' my.address <- cartociudad_geocode(full_address = address)
```
Misma recomendación que con `cartociudad_reverse_geocode`: sugiero incluir un test para múltiples direcciones

koldLight commented on DESCRIPTION:
```diff
@@ -9,7 +9,7 @@ Authors@R: c(person("Carlos J.", "Gil Bellosta", email="cgb@datanalytics.com", r
 Maintainer: Carlos J. Gil Bellosta <cgb@datanalytics.com>
 Description: Access to Cartociudad cartography API, which provides mapping and other related services for Spain.
-Imports: httr, jsonlite, xml2, plyr, geosphere
+Imports: httr, jsonlite, xml2, plyr, geosphere, purrr
```
No sé si el uso actual justifica incluir una nueva dependencia (`purrr`). Actualmente solo se utiliza para hacer 2x rbind de lista de data.frames, que se puede sustituir por funciones de R base: `do.call(rbind, res_list)`

carlosvergara replied:
Sí, en principio lo hice así. He de reconocer que soy un adicto al paquete `purrr`... :) Lo corrijo.

carlosvergara replied:
He pensado en usar `plyr::rbind.fill()`, puesto que los `data.frame`'s pueden tener diferente número de columnas.

koldLight replied:
Genial

koldLight reviewed (changes requested):
Me parece una mejora muy importante la vectorización de las funciones, para poder hacer `cartociudad_reverse_geocode(mi_df$latitude, mi_df$longitude)` en lugar de hacer el bucle por fuera.

He añadido algunas sugerencias de cambio a la PR, comenta sobre ellas si tienes otra opinión.

Gracias por colaborar en el desarrollo del paquete, son buenos y grandes cambios :)

carlosvergara committed:
Change version to 0.5.3

- Remove purrr from Imports: purrr:map_df to plyr::rbind.fill.

- Original names in results from cartociudad_reverse_geocode.

carlosvergara committed:
New tests & old names to reverse geocoding results

carlosvergara commented:
Cambios realizados. He subido a versión 0.5.3

koldLight reviewed (approved):
Genial el repaso a los métodos utilizados en los tests y el cambio a otros más legibles

Commit:
Update version to 0.5.4

Add progress bar to cartociudad_geocode & cartociudad_reverse_geocode.
Add utils to Imports.
Add new argument to cartociudad_geocode to use the previous geocoder version and adapt code to it.

cjgb reviewed (changes requested):
Tengo que probar el nuevo código que mandas. De todos modos, ¿por qué eliminas la referencia catastral (y otras partes de la salida)? Es muy útil en muchas aplicaciones.

Te comento también que vamos a hacer cambios en la manera de organizar la gestión de este y otros paquetes similares pronto. Por eso hemos tardado en responder. Dado tu interés, te tendremos informado.

carlosvergara committed:
Server error handling

carlosvergara committed:
Fix merge conflict

carlosvergara committed:
Fix "server connection" bug in geocoding functions

carlosvergara committed:
Fix "server connection" bug in geocoding functions

carlosvergara committed:
Merge conflict
